test/test_pid_reuse.c

```diff
@@ -0,0 +1,172 @@
+ *
+ * GPLv2:
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ *
+ * Apache v2.0:
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/syscall.h>
+#include <sys/wait.h>
+#include <assert.h>
+#include <errno.h>
+#include <pthread.h>
+#include "zlib.h"
+
+/* Globals shared by all threads. */
+z_stream c_stream;
+
+void* comp_file_multith (void *iterations)
+{
+	int err;
+	long *sleep_sec;
+
+	pid_t my_pid = syscall(SYS_gettid);
+	fprintf(stderr, "thread started: pid: %d\n", (int) my_pid);
+
+	/* Open the window. */
+	err = deflateInit(&c_stream, Z_DEFAULT_COMPRESSION);
+	if (err != 0) {
+		printf("nx_deflateInit err %d\n", err);
+		return (void *) 1;
+	}
+
+	sleep_sec = (long *) iterations;
+	sleep(sleep_sec[0]);
+
+	fprintf(stderr, "thread %d exits\n", (int) my_pid);
+	return (void *) 0;
+}
+
+void fork_children (int parent_pid) {
+	int child_id, i = 0;
+	int seconds = 10;
+	int children = 2048;
+
+	for (i=0; i < children; i++) {
+		child_id = fork();
+
+		if (child_id > 0) {
+			/* PID is being reused. */
+			if (child_id == parent_pid) {
+				fprintf(stderr, "!!! pid %d reused (see ps)\n",
+					child_id);
+				/* Keep this child alive enough time to hit the
+				 * csb error signal. */
+				sleep(300);
+			}
+			continue;
+		} else if (child_id == 0) {
+			pid_t pid = getpid();
+			fprintf(stderr, "%d: child(%d)\n", i, (int) pid);
+			/* Wait enough time to fork all children. */
+			sleep(seconds);
+			fprintf(stderr, "%d: child(%d) exits\n", i, (int) pid);
+			exit(0);
+		} else if (child_id == -1) {
+			fprintf(stderr, "fork failed: %s\n", strerror(errno));
+			exit(-1);
+		}
+	}
+}
+
+int main(int argc, char **argv)
+{
+	int rc, compLen = 1024;
+	char *compr;
+	pthread_t thread;
+	void *ret;
+	long iterations;
+	z_const char hello[] = "hello, hello!";
+
+	if (argc == 2)
+		iterations = atoi(argv[1]);
+	else
+		iterations = 100;
+
+	/* Get parent thread kernel PID. */
```

Comment:
Should this read "Get parent thread ID"?  Similarly renaming the variable.

Reply:
@pmur, we are looking for a kernel problem, I believe it's better to use the kernel naming and call it PID instead of TID.